File: demos/typescript/geometry-samples/src/samples/raycast.ts
// Raycasting a triangle mesh.
//
// `Triangle3D.Raycast(Ray3D)` is the stdlib's Moller-Trumbore intersection: it
// solves the ray/triangle system for the barycentric pair and the distance in
// one pass, with no precomputed plane, and returns the full RayHit3D record
// (distance, position, ray-facing normal, barycentric coordinate, UV).
//
// The sweep over the mesh's triangles is brute force. `TriangleMesh3D.Primitives`
// gathers the faces into concrete Triangle3D values; the BVH sample is the
// version that prunes.

import type { Drawable, Sample } from '../core/types.js';
import {
    Direction3D, Line3D, Point3D, Ray3D, RayHit3D, TriangleMesh3D, Vector3D,
} from '../plato/plato.g.js';
import { toArray } from '../core/meshBuilder.js';
import { buildIcosphere } from './icosphere.js';

export interface MeshHit {
    hit: RayHit3D;
    triangle: number;
}

/** Closest hit against every triangle of the mesh. */
export function raycastMesh(mesh: TriangleMesh3D, ray: Ray3D): MeshHit | null {
    let best: MeshHit | null = null;
    const triangles = toArray(mesh.Primitives());
    for (let i = 0; i < triangles.length; i++) {
        const hit = triangles[i].Raycast(ray);
        if (hit.Hit && (best === null || hit.Distance < best.hit.Distance))
            best = { hit, triangle: i };
    }
    return best;
}

export const raycastSample: Sample = {
    id: 'raycast',
    title: 'Raycasting',
    description: 'Triangle3D.Raycast (Moller-Trumbore) swept over TriangleMesh3D.Primitives.',
    build(): Drawable[] {
        const mesh = buildIcosphere(2);
        const direction = new Direction3D(new Vector3D(0, 0, -1));

        const rays: Line3D[] = [];
        const misses: Line3D[] = [];
        const hits: Point3D[] = [];
        for (let iy = 0; iy < 9; iy++) {
            for (let ix = 0; ix < 9; ix++) {
                const origin = new Point3D((ix / 8 - 0.5) * 2.6, (iy / 8 - 0.5) * 2.6, 2.2);
                const ray = new Ray3D(origin, direction);
                const found = raycastMesh(mesh, ray);
                if (found) {
                    rays.push(new Line3D(origin, found.hit.Position));
                    hits.push(found.hit.Position);
                } else {
                    misses.push(new Line3D(origin, ray.PointAt(4.4)));
                }
            }
        }

        return [
            { kind: 'mesh', mesh, color: 0x4da3ff, opacity: 0.45 },
            { kind: 'lines', segments: rays, color: 0xffd166, opacity: 0.9 },
            { kind: 'lines', segments: misses, color: 0x39414e, opacity: 0.5 },
            { kind: 'points', points: hits, color: 0xff6b6b, size: 0.07 },
        ];
    },
};

File: demos/rust/geometry-samples/src/samples/raycast.rs
// Ray-triangle intersection (Moller-Trumbore).
//
// Solves the ray/triangle equation directly with two cross products, giving
// the barycentric coordinates (u, v) and the distance t in one pass, with no
// precomputed plane. Written fluently against the Plato vector library -
// compare with the C# and TypeScript versions, which read identically.

use crate::core::mesh_builder::{push_vector, vertex_at};
use crate::core::types::{Drawable, MeshData};
use crate::plato::*;

use super::icosphere::build_icosphere;

pub struct RayHit {
    pub t: f64,
    pub point: Vector3D,
    pub triangle: usize,
}

pub fn intersect_triangle(
    origin: Vector3D,
    dir: Vector3D,
    a: Vector3D,
    b: Vector3D,
    c: Vector3D,
) -> Option<f64> {
    let e1 = b.Subtract(a);
    let e2 = c.Subtract(a);
    let p = dir.Cross(e2);
    let det = e1.Dot(p);
    if det.Abs() < 1e-10 {
        return None; // parallel to the triangle plane
    }
    let inv = 1.0 / det;
    let s = origin.Subtract(a);
    let u = s.Dot(p) * inv;
    if !(0.0..=1.0).contains(&u) {
        return None;
    }
    let q = s.Cross(e1);
    let v = dir.Dot(q) * inv;
    if v < 0.0 || u + v > 1.0 {
        return None;
    }
    let t = e2.Dot(q) * inv;
    if t > 1e-9 {
        Some(t)
    } else {
        None
    }
}

/// Closest hit against every triangle (brute force; see the BVH sample).
pub fn raycast_mesh(mesh: &MeshData, origin: Vector3D, dir: Vector3D) -> Option<RayHit> {
    let mut best: Option<RayHit> = None;
    for tri in 0..mesh.indices.len() / 3 {
        let a = vertex_at(&mesh.positions, mesh.indices[tri * 3] as usize);
        let b = vertex_at(&mesh.positions, mesh.indices[tri * 3 + 1] as usize);
        let c = vertex_at(&mesh.positions, mesh.indices[tri * 3 + 2] as usize);
        if let Some(t) = intersect_triangle(origin, dir, a, b, c) {
            if best.as_ref().is_none_or(|h| t < h.t) {
                best = Some(RayHit { t, point: origin.Add(dir.Scale(t)), triangle: tri });
            }
        }
    }
    best
}

pub fn build() -> Vec<Drawable> {
    let mesh = build_icosphere(2);

    let mut rays: Vec<f64> = Vec::new();
    let mut hits: Vec<f64> = Vec::new();
    let mut misses: Vec<f64> = Vec::new();
    let dir = Vector3D::new(0.0, 0.0, -1.0);
    for iy in 0..9 {
        for ix in 0..9 {
            let origin = Vector3D::new(
                (ix as f64 / 8.0 - 0.5) * 2.6,
                (iy as f64 / 8.0 - 0.5) * 2.6,
                2.2,
            );
            match raycast_mesh(&mesh, origin, dir) {
                Some(hit) => {
                    push_vector(&mut rays, origin);
                    push_vector(&mut rays, hit.point);
                    push_vector(&mut hits, hit.point);
                }
                None => {
                    push_vector(&mut misses, origin);
                    push_vector(&mut misses, origin.Add(dir.Scale(4.4)));
                }
            }
        }
    }

    vec![
        mesh.into_drawable(),
        Drawable::Lines { positions: rays },
        Drawable::Lines { positions: misses },
        Drawable::Points { positions: hits },
    ]
}
